src/server/webkit.ts

```diff
@@ -236,36 +236,149 @@ function getMacVersion(): string {
+      if (!value)
+        return;
+      const { id, socket } = value;
+
+      if (pendingBrowserContextCreations.has(parsedMessage.id)) {
+        // Browser.createContext response -> establish context attribution.
+        browserContextIds.set(parsedMessage.result.browserContextId, socket);
+        pendingBrowserContextCreations.delete(parsedMessage.id);
+      }
+
+      const deletedContextId = pendingBrowserContextDeletions.get(parsedMessage.id);
+      if (deletedContextId) {
+        // Browser.createContext response -> remove context attribution.
+        browserContextIds.delete(deletedContextId);
+        pendingBrowserContextDeletions.delete(parsedMessage.id);
+      }
+
+      parsedMessage.id = id;
+      socket.send(JSON.stringify(parsedMessage));
       return;
     }
-    if (socket) {
-      s.close(undefined, 'Multiple connections are not supported');
+
+    // Process notification response.
+    const { method, params, pageProxyId } = parsedMessage;
+    if (pageProxyId) {
+      const socket = pageProxyIds.get(pageProxyId);
+      if (!socket || socket.readyState === ws.CLOSING) {
+        // Drop unattributed messages on the floor.
+        return;
+      }
+      socket.send(message);
       return;
     }
-    socket = s;
-    s.on('message', message => transport.send(Buffer.from(message).toString()));
-    transport.onmessage = message => {
-      // We are not notified when socket starts closing, and sending messages to a closing
-      // socket throws an error.
-      if (s.readyState !== ws.CLOSING)
-        s.send(message);
-    };
-    s.on('close', () => {
-      socket = undefined;
-      transport.onmessage = undefined;
+    if (method === 'Browser.pageProxyCreated') {
+      const socket = browserContextIds.get(params.pageProxyInfo.browserContextId);
+      if (!socket || socket.readyState === ws.CLOSING) {
+        // Drop unattributed messages on the floor.
+        return;
+      }
+      pageProxyIds.set(params.pageProxyInfo.pageProxyId, socket);
+      socket.send(message);
+      return;
+    }
+    if (method === 'Browser.pageProxyDestroyed') {
+      const socket = pageProxyIds.get(params.pageProxyId);
+      pageProxyIds.delete(params.pageProxyId);
+      if (socket && socket.readyState !== ws.CLOSING)
+        socket.send(message);
+      return;
+    }
+    if (method === 'Browser.provisionalLoadFailed') {
+      const socket = pageProxyIds.get(params.pageProxyId);
+      if (socket && socket.readyState !== ws.CLOSING)
+        socket!.send(message);
+      return;
+    }
+  };
+
+  server.on('connection', (socket: ws, req) => {
+    if (req.url !== '/' + guid) {
+      socket.close();
+      return;
+    }
+    sockets.add(socket);
+    // Following two messages are reporting the default browser context and the default page.
+    socket.send(JSON.stringify({
+      method: 'Browser.pageProxyCreated',
+      params: { pageProxyInfo: { pageProxyId: '5', browserContextId: '0000000000000002' } }
```

Comment:
Can we simply wait for the first pageProxyCreated/targetCreated and broadcast it?

Reply:
Follow up!